File: src/article_analysis/protocol/types-test.cc
#include <string>

#include <gtest/gtest.h>

#include "protocol/types.h"


using std::string;


namespace {


TEST(DocIdentity, All) {
  string s("\x01\x00\x00\x00\x00\x00\x00\x00""a"
           "\x03\x00\x00\x00", 8 + 1 + 4);
  protocol::types::DocIdentity di;
  size_t offs = 0;
  bool ret = di.Load(s, &offs);
  EXPECT_TRUE(ret);
  EXPECT_EQ(offs, 8u + 1 + 4);
  EXPECT_EQ(di.board, "a");
  EXPECT_EQ(di.id, 3);
}


TEST(ReplyMessage, Load) {
  protocol::types::ReplyMessage rm;

  string s("\x01\x00"
           "\x03\x00\x00\x00\x00\x00\x00\x00""abc"
           "\x02\x00\x00\x00\x00\x00\x00\x00""ab",
           2 + 8 + 3 + 8 + 2);
  size_t offs = 0;
  bool ret = rm.Load(s, &offs);
  EXPECT_TRUE(ret);
  EXPECT_EQ(offs, s.length());
  EXPECT_EQ(rm.mode, protocol::types::ReplyMode::NORMAL);
  EXPECT_EQ(rm.user, "abc");
  EXPECT_EQ(rm.message, "ab");
}


TEST(DocMetaData, Load) {
  protocol::types::DocMetaData dm;
  string s("\x02\x00\x00\x00"
           "\x01\x00\x00\x00"
           "\x08\x00\x00\x00\x00\x00\x00\x00""abcdmeow"
           "\x03\x00\x00\x00\x00\x00\x00\x00""xyz"
           "\x04\x00\x00\x00\x00\x00\x00\x00"
           "\x05\x00\x00\x00\x00\x00\x00\x00""12345"
           "\x0a\x00\x00\x00"
           "\x0b\x00\x00\x00"
           "\x0c\x00\x00\x00",
           4 + 4 + (8 + 8) + (8 + 3) + 8 + (8 + 5) + 4 + 4 + 4);
  size_t offs = 0;
  bool ret = dm.Load(s, &offs);
  EXPECT_TRUE(ret);
  EXPECT_EQ(offs, s.length());
  EXPECT_EQ(dm.id, 2);
  EXPECT_EQ(dm.prev_id, 1);
  EXPECT_EQ(dm.title, "abcdmeow");
  EXPECT_EQ(dm.author, "xyz");
  EXPECT_EQ(dm.post_time, 4);
  EXPECT_EQ(dm.board, "12345");
  EXPECT_EQ(dm.num_reply_rows[0], 10u);
  EXPECT_EQ(dm.num_reply_rows[1], 11u);
  EXPECT_EQ(dm.num_reply_rows[2], 12u);
}


TEST(DocRealData, Load) {
  protocol::types::DocRealData dd;
  string s("\x07\x00\x00\x00\x00\x00\x00\x00""1234567"
           "\x01\x00\x00\x00\x00\x00\x00\x00"
           "\x01\x00"
           "\x03\x00\x00\x00\x00\x00\x00\x00""abc"
           "\x02\x00\x00\x00\x00\x00\x00\x00""ab",
           (8 + 7) + 8 + 2 + (8 + 3) + (8 + 2));
  size_t offs = 0;
  bool ret = dd.Load(s, &offs);
  EXPECT_TRUE(ret);
  EXPECT_EQ(offs, s.length());
  EXPECT_EQ(dd.content, "1234567");
  ASSERT_EQ(dd.reply_messages.size(), 1u);
  EXPECT_EQ(dd.reply_messages[0].mode, protocol::types::ReplyMode::NORMAL);
  EXPECT_EQ(dd.reply_messages[0].user, "abc");
  EXPECT_EQ(dd.reply_messages[0].message, "ab");
}

}  // namespace

File: src/miner/modules/protocol/types_test.py
import inspect
import sys
import unittest

from modules.protocol import types


def load_tests(loader, tests, pattern):
    suite = unittest.TestSuite()
    for (name, obj) in inspect.getmembers(sys.modules[__name__]):
        if name.startswith('Test'):
            tests = loader.loadTestsFromTestCase(obj)
            suite.addTests(tests)
    return suite


class TestReplyMessage(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestReplyMessage, self).__init__(*args, **kwargs)

    def runTest(self):
        rm = types.ReplyMessage(types.ReplyMode.NORMAL, 'meow', 'wang wang')
        buf = b''.join([b'\x01\x00',
                        b'\x04\x00\x00\x00\x00\x00\x00\x00meow',
                        b'\x09\x00\x00\x00\x00\x00\x00\x00wang wang'])
        self.assertEqual(rm.dump(), buf)


class TestDocMetaData(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestDocMetaData, self).__init__(*args, **kwargs)

    def runTest(self):
        dmd = types.DocMetaData(5, 3, 'meow', 'wang', 9, 'blah', [2, 5, 7])
        buf = b''.join([b'\x05\x00\x00\x00',
                        b'\x03\x00\x00\x00',
                        b'\x04\x00\x00\x00\x00\x00\x00\x00meow',
                        b'\x04\x00\x00\x00\x00\x00\x00\x00wang',
                        b'\x09\x00\x00\x00\x00\x00\x00\x00',
                        b'\x04\x00\x00\x00\x00\x00\x00\x00blah',
                        b'\x02\x00\x00\x00',
                        b'\x05\x00\x00\x00',
                        b'\x07\x00\x00\x00'])
        print('')
        print(dmd.dump())
        print(buf)
        self.assertEqual(dmd.dump(), buf)


class TestDocIdentity(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestDocIdentity, self).__init__(*args, **kwargs)

    def runTest(self):
        di = types.DocIdentity('blabla~', 18)
        buf = b''.join([b'\x07\x00\x00\x00\x00\x00\x00\x00blabla~',
                        b'\x12\x00\x00\x00'])
        self.assertEqual(di.dump(), buf)

        (di, offs) = types.DocIdentity.load(buf)
        self.assertEqual(offs, 8 + 7 + 4)
        self.assertEqual(di.board, 'blabla~')
        self.assertEqual(di.idid, 18)


class TestDocRealData(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestDocRealData, self).__init__(*args, **kwargs)

    def runTest(self):
        r1 = types.ReplyMessage(2, 'a', 'bb')
        r2 = types.ReplyMessage(1, 'b', 'cc')
        drd = types.DocRealData('wu lala', [r1, r2])
        buf = b''.join([b'\x07\x00\x00\x00\x00\x00\x00\x00wu lala',
                        b'\x02\x00\x00\x00\x00\x00\x00\x00',
                        r1.dump(), r2.dump()])
        self.assertEqual(drd.dump(), buf)


class TestDocRelInfo(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestDocRelInfo, self).__init__(*args, **kwargs)

    def runTest(self):
        dri = types.DocRelInfo([], [], [types.DocIdentity('board', 123)])
        buf = b'\x00\x00\x00\x00\x00\x00\x00\x00' + \
                b'\x00\x00\x00\x00\x00\x00\x00\x00' + \
                b'\x01\x00\x00\x00\x00\x00\x00\x00' + \
                types.DocIdentity('board', 123).dump()

        buf2 = dri.dump()
        self.assertEqual(buf, buf2)

        dri2 = types.DocRelInfo.load(buf)[0]
        self.assertEqual(len(dri2.pos_rel_docs), 0)
        self.assertEqual(len(dri2.neg_rel_docs), 0)
        self.assertEqual(len(dri2.neutral_rel_docs), 1)
        self.assertEqual(dri2.neutral_rel_docs[0].board, 'board')
        self.assertEqual(dri2.neutral_rel_docs[0].idid, 123)
